File: libs/moqtail-ts/src/model/control/track_status_request.ts
import { BaseByteBuffer, ByteBuffer, FrozenByteBuffer } from '../common/byte_buffer'
import { Tuple } from '../common/tuple'
import { KeyValuePair } from '../common/pair'
import { ControlMessageType } from './constant'
import { LengthExceedsMaxError } from '../error/error'
import { FullTrackName } from '../data'

export class TrackStatusRequest {
  constructor(
    public readonly requestId: bigint,
    public readonly fullTrackName: FullTrackName,
    public readonly parameters: KeyValuePair[],
  ) {}

  getType(): ControlMessageType {
    return ControlMessageType.TrackStatusRequest
  }

  serialize(): FrozenByteBuffer {
    const buf = new ByteBuffer()
    buf.putVI(ControlMessageType.TrackStatusRequest)
    const payload = new ByteBuffer()
    payload.putVI(this.requestId)
    payload.putFullTrackName(this.fullTrackName)
    payload.putVI(this.parameters.length)
    for (const param of this.parameters) {
      payload.putKeyValuePair(param)
    }
    const payloadBytes = payload.toUint8Array()
    if (payloadBytes.length > 0xffff) {
      throw new LengthExceedsMaxError('TrackStatusRequest::serialize(payloadBytes.length)', 0xffff, payloadBytes.length)
    }
    buf.putU16(payloadBytes.length)
    buf.putBytes(payloadBytes)
    return buf.freeze()
  }

  static parsePayload(buf: BaseByteBuffer): TrackStatusRequest {
    const requestId = buf.getVI()
    const fullTrackName = buf.getFullTrackName()
    const paramCount = buf.getNumberVI()
    const parameters: KeyValuePair[] = new Array(paramCount)
    for (let i = 0; i < paramCount; i++) {
      parameters[i] = buf.getKeyValuePair()
    }
    return new TrackStatusRequest(requestId, fullTrackName, parameters)
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest
  describe('TrackStatusRequest', () => {
    test('roundtrip', () => {
      const requestId = 241421n
      const trackNamespace = Tuple.fromUtf8Path('charlie/chocolate/factory')
      const trackName = new TextEncoder().encode('OompaLumpa')
      const fullTrackName = FullTrackName.tryNew(trackNamespace, trackName)
      const parameters = [
        KeyValuePair.tryNewVarInt(0, 10),
        KeyValuePair.tryNewBytes(1, new TextEncoder().encode('Chocomocco?!')),
      ]
      const msg = new TrackStatusRequest(requestId, fullTrackName, parameters)
      const frozen = msg.serialize()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.TrackStatusRequest))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining)
      const deserialized = TrackStatusRequest.parsePayload(frozen)
      expect(deserialized).toEqual(msg)
      expect(frozen.remaining).toBe(0)
    })
    test('excess roundtrip', () => {
      const requestId = 241421n
      const trackNamespace = Tuple.fromUtf8Path('charlie/chocolate/factory')
      const trackName = new TextEncoder().encode('OompaLumpa')
      const fullTrackName = FullTrackName.tryNew(trackNamespace, trackName)
      const parameters = [
        KeyValuePair.tryNewVarInt(0, 10),
        KeyValuePair.tryNewBytes(1, new TextEncoder().encode('Chocomocco?!')),
      ]
      const msg = new TrackStatusRequest(requestId, fullTrackName, parameters)
      const serialized = msg.serialize().toUint8Array()
      const excess = new Uint8Array([9, 1, 1])
      const buf = new ByteBuffer()
      buf.putBytes(serialized)
      buf.putBytes(excess)
      const frozen = buf.freeze()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.TrackStatusRequest))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining - 3)
      const deserialized = TrackStatusRequest.parsePayload(frozen)
      expect(deserialized).toEqual(msg)
      expect(frozen.remaining).toBe(3)
      expect(Array.from(frozen.getBytes(3))).toEqual([9, 1, 1])
    })
    test('partial message', () => {
      const requestId = 241421n
      const trackNamespace = Tuple.fromUtf8Path('charlie/chocolate/factory')
      const trackName = new TextEncoder().encode('OompaLumpa')
      const fullTrackName = FullTrackName.tryNew(trackNamespace, trackName)
      const parameters = [
        KeyValuePair.tryNewVarInt(0, 10),
        KeyValuePair.tryNewBytes(1, new TextEncoder().encode('Chocomocco?!')),
      ]
      const msg = new TrackStatusRequest(requestId, fullTrackName, parameters)
      const serialized = msg.serialize().toUint8Array()
      const upper = Math.floor(serialized.length / 2)
      const partial = serialized.slice(0, upper)
      const frozen = new FrozenByteBuffer(partial)
      expect(() => {
        frozen.getVI()
        frozen.getU16()
        TrackStatusRequest.parsePayload(frozen)
      }).toThrow()
    })
  })
}

File: libs/moqtail-rs/src/model/control/track_status_request.rs
use super::constant::ControlMessageType;
use super::control_message::ControlMessageTrait;
use crate::model::common::pair::KeyValuePair;
use crate::model::common::tuple::Tuple;
use crate::model::common::varint::{BufMutVarIntExt, BufVarIntExt};
use crate::model::error::ParseError;
use bytes::{Buf, BufMut, Bytes, BytesMut};

#[derive(Debug, PartialEq, Clone)]
pub struct TrackStatusRequest {
  pub request_id: u64,
  pub track_namespace: Tuple,
  pub track_name: String,
  pub parameters: Vec<KeyValuePair>,
}

impl TrackStatusRequest {
  pub fn new(
    request_id: u64,
    track_namespace: Tuple,
    track_name: String,
    parameters: Vec<KeyValuePair>,
  ) -> Self {
    Self {
      request_id,
      track_namespace,
      track_name,
      parameters,
    }
  }
}

impl ControlMessageTrait for TrackStatusRequest {
  fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut buf = BytesMut::new();
    buf.put_vi(ControlMessageType::TrackStatusRequest)?;

    let mut payload = BytesMut::new();
    payload.put_vi(self.request_id)?;
    payload.extend_from_slice(&self.track_namespace.serialize()?);
    payload.put_vi(self.track_name.len())?;
    payload.extend_from_slice(self.track_name.as_bytes());
    payload.put_vi(self.parameters.len())?;
    for param in &self.parameters {
      payload.extend_from_slice(&param.serialize()?);
    }

    let payload_len: u16 = payload
      .len()
      .try_into()
      .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
        context: "TrackStatusRequest::serialize(payload_length)",
        from_type: "usize",
        to_type: "u16",
        details: e.to_string(),
      })?;

    buf.put_u16(payload_len);
    buf.extend_from_slice(&payload);

    Ok(buf.freeze())
  }

  fn parse_payload(payload: &mut Bytes) -> Result<Box<Self>, ParseError> {
    let request_id = payload.get_vi()?;
    let track_namespace = Tuple::deserialize(payload)?;
    let name_length = payload.get_vi()?;
    let name_length: usize = name_length
      .try_into()
      .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
        context: "TrackStatusRequest::parse_payload(name_length)",
        from_type: "u64",
        to_type: "usize",
        details: e.to_string(),
      })?;

    if payload.remaining() < name_length {
      return Err(ParseError::NotEnoughBytes {
        context: "TrackStatusRequest::parse_payload(name_length)",
        needed: name_length,
        available: payload.remaining(),
      });
    }
    let name_bytes = payload.copy_to_bytes(name_length);

    let track_name =
      String::from_utf8(name_bytes.to_vec()).map_err(|e| ParseError::InvalidUTF8 {
        context: "TrackStatusRequest::parse_payload(track_name)",
        details: e.to_string(),
      })?;

    let param_count_u64 = payload.get_vi()?;
    let param_count: usize =
      param_count_u64
        .try_into()
        .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
          context: "Subscribe::deserialize(param_count)",
          from_type: "u64",
          to_type: "usize",
          details: e.to_string(),
        })?;

    let mut parameters = Vec::with_capacity(param_count);
    for _ in 0..param_count {
      let param = KeyValuePair::deserialize(payload)?;
      parameters.push(param);
    }

    Ok(Box::new(TrackStatusRequest {
      request_id,
      track_namespace,
      track_name,
      parameters,
    }))
  }

  fn get_type(&self) -> ControlMessageType {
    ControlMessageType::TrackStatusRequest
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use bytes::Buf;

  #[test]
  fn test_roundtrip() {
    let request_id = 241421;
    let track_namespace = Tuple::from_utf8_path("charlie/chocolate/factory");
    let track_name = "OompaLumpa".to_string();
    let parameters = vec![
      KeyValuePair::try_new_varint(0, 10).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"Chocomocco?!")).unwrap(),
    ];
    let track_status_request = TrackStatusRequest {
      request_id,
      track_namespace,
      track_name,
      parameters,
    };

    let mut buf = track_status_request.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::TrackStatusRequest as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = TrackStatusRequest::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, track_status_request);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_excess_roundtrip() {
    let request_id = 241421;
    let track_namespace = Tuple::from_utf8_path("charlie/chocolate/factory");
    let track_name = "OompaLumpa".to_string();
    let parameters = vec![
      KeyValuePair::try_new_varint(0, 10).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"Chocomocco?!")).unwrap(),
    ];
    let track_status_request = TrackStatusRequest {
      request_id,
      track_namespace,
      track_name,
      parameters,
    };

    let serialized = track_status_request.serialize().unwrap();
    let mut excess = BytesMut::new();
    excess.extend_from_slice(&serialized);
    excess.extend_from_slice(&[9u8, 1u8, 1u8]);
    let mut buf = excess.freeze();

    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::TrackStatusRequest as u64);
    let msg_length = buf.get_u16();

    assert_eq!(msg_length as usize, buf.remaining() - 3);
    let deserialized = TrackStatusRequest::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, track_status_request);
    assert_eq!(buf.chunk(), &[9u8, 1u8, 1u8]);
  }

  #[test]
  fn test_partial_message() {
    let request_id = 241421;
    let track_namespace = Tuple::from_utf8_path("charlie/chocolate/factory");
    let track_name = "OompaLumpa".to_string();
    let parameters = vec![
      KeyValuePair::try_new_varint(0, 10).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"Chocomocco?!")).unwrap(),
    ];
    let track_status_request = TrackStatusRequest {
      request_id,
      track_namespace,
      track_name,
      parameters,
    };

    let mut buf = track_status_request.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::TrackStatusRequest as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());

    let upper = buf.remaining() / 2;
    let mut partial = buf.slice(..upper);
    let deserialized = TrackStatusRequest::parse_payload(&mut partial);
    assert!(deserialized.is_err());
  }
}
